rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,221 @@
+
+Further, while a full network partition between two groups of nodes will result in a "split brain" scenario in an unavoidable fashion, with LogRendezvousHash it is possible to counteract a partial split by way of a request forwarding mechanism. If a node cannot directly reach another node, it should try to ask all other nodes in its list if any of them can reach the node. If so, the first one to respond in the positive becomes the new "router" that all requests to that node are forwarded through, with periodic attempts in the background to remove the extra hop by re-polling. Only if all nodes a node can reach report that they also cannot reach the node in question should it be marked as "down" and assignment of keys shuffled with the new, reduced set. This is fundamentally how the internet routes packets to unknown MAC addresses, and can guard against misconfigured firewalls/routers black hole-ing packets that they shouldn't be, but only works when there is an unaffected third party, so requires at least three separate sub-clusters to work.
+
+For example, let's presume a cluster has been deployed to all three major cloud providers: AWS, Azure, and GCP. Now let's presume that GCP has a network configuration error that makes AWS nodes unreachable to it and vice versa. When the GCP nodes want to get data that was previously assigned to AWS, they will fail to contact the AWS node, then try to ask all other nodes (which would include the other AWS nodes) which ones can reach the AWS node in question. The GCP nodes would report a failure, but the Azure nodes would report a success, and one of those nodes will be chosen effectively at random to be the node that will route to the AWS node. This process would repeat across every GCP node and AWS node combination with an effectively random assignment of Azure nodes to handle the load, which should lead to an approximately even level of traffic distribution between the Azure nodes.
+
+![A partial partition diagram of the paragraph above](./partial_partition.jpg)
+
+This all can be contained in a relatively simple `/ping/<public_ip_address>` control port endpoint. Normally the `/ping/<public_ip_address>` would be directed at `<public_ip_address>`, which should always return `ok` (as long as the cluster secret token is also provided as a header, otherwise `fail`, but that's at an earlier part of the control port implementation). But if a request for `/ping/<public_ip_address>` comes in and `<public_ip_address>` is not the node's own public IP address, then it immediately forwards the request on to the public IP address in the request URL and returns its response or `fail` if it times out (should be a short timeout as well, such as 50ms, but exact threshold depends on real world data here).
```

Comment:
Will there any way to know if a request was forwarded/proxied? If not, why doesn't it matter?

Reply:
typically (reverse) proxies use `X-forwarded-for` - we can probably use that. I imagine we'll have to return the status of the daemon at the address

Reply:
So if you call `/ping/<someIPaddress>` to `someIPaddress` you know there's no forwarding because you're pinging that IP address directly. If you call `/ping/<someIPaddress>` to `someOtherIPAddress` then you *know* that there's forwarding going on because you're explicitly asking for it to be forwarded?

Was this question meant for a different paragraph, maybe? The other endpoints could silently forward for you, but `/ping` is pretty simple, I thought.

Reply:
The question was in regards to all cluster coordination. I like that it is as simple as looking at the ip address in the url and comparing it to itself.

Reply:
So we can use `X-forwarded-for` as @cdmistman suggested for easier debugging, but from the API perspective the automatic forwarding was to abstract the routing logic from higher layers in the stack (and from the Anycloud apps). It also forced a more complete solution because all possible conditions of a smaller surface area were considered (all nodes up, one or more nodes actually down, partial partitioning of the cluster, and full partitioning of the cluster into two or more sub-clusters, where the latter was decided to be an unresolveable scenario, but *every other outage scenario* is now recoverable :) ).